File: Leetcode/Python3/20. Valid Parentheses/solution.py
# https://leetcode.com/problems/valid-parentheses/


class Solution:
    def isValid(self, s: str) -> bool:
        parenthesis = {")": "(", "}": "{", "]": "["}
        stack = []
        for p in s:
            if p not in parenthesis:  # (, {, [
                stack.append(p)
            elif not stack or stack.pop() != parenthesis[p]:
                return False
        return len(stack) == 0  # If stack not empty there is single open parenthesis

File: Leetcode/C++/20. Valid Parentheses/solution.cpp
// https://leetcode.com/problems/valid-parentheses/

#include <vector>
#include <map>
#include <string>

using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        vector<int> stack;
        unordered_map<char, char> parenthesis = {
            {')', '('},
            {'}', '{'},
            {']', '['}};
        for (char c : s)
        {
            if (parenthesis.find(c) == parenthesis.end())
            {
                stack.push_back(c);
            }
            else
            {
                // If parenthesis invaild return false
                if (stack.empty() || stack.back() != parenthesis[c])
                {
                    return false;
                }
                stack.pop_back();
            }
        }
        // If stack is not empty invalid parenthesis
        return stack.empty();
    }
};